### PR TITLE
Integrate trace headers

### PR DESCRIPTION
**Pull request by Tyrrrz**
Related to #663 

Note, this doesn't include auto instrumentation for ASP.net Core, but only the core parts. Adding the instrumentation later would be easy once these are in (and won't be a breaking change).

### CODE REVIEW

File: src/Sentry/Internal/Hub.cs [hunk 1]
```diff
@@ -158,6 +158,7 @@ public void WithScope(Action<Scope> scopeCallback)
 
         public SentryTraceHeader? GetSentryTrace()
         {
+            // TODO:
             var (currentScope, _) = ScopeManager.GetCurrent();
             return currentScope.Transaction?.GetTraceHeader();
         }
```
Comment (Tyrrrz) [inline src/Sentry/Internal/Hub.cs:164]: @bruno-garcia this previously relied that transaction is added to the current scope.
How should we get current transaction now? Or is it current span? Can you please explain how `Hub.GetSentryTrace()` should work?

File: src/Sentry/Protocol/SentryTraceHeader.cs [hunk 2]
```diff
@@ -7,24 +8,59 @@ namespace Sentry.Protocol
...
+        /// <summary>
+        /// Injects trace information into the default headers of the specified HTTP client.
+        /// </summary>
+        public void Inject(HttpClient client)
+        {
+            var headerValue = ToString();
+
+            client.DefaultRequestHeaders.Remove(HeaderName);
+            client.DefaultRequestHeaders.Add(HeaderName, headerValue);
         }
```
Comment (Tyrrrz) [inline src/Sentry/Protocol/SentryTraceHeader.cs:58]: It's either instance methods on `SentryTraceHeaders` or extension methods on `HttpRequestMessage` and `HttpClient` respectively. I have no preference. Thoughts?

File: src/Sentry/Protocol/SentryTraceHeader.cs [hunk 3]
```diff
@@ -7,24 +10,60 @@ namespace Sentry.Protocol
...
+        /// Injects trace information into the headers of the specified HTTP request.
+        /// </summary>
+        public void Inject(HttpRequestMessage request) =>
+            Inject(request.Headers);
+
+        /// <summary>
+        /// Injects trace information into the default headers of the specified HTTP client.
+        /// </summary>
+        public void Inject(HttpClient client) =>
+            Inject(client.DefaultRequestHeaders);
```
Comment (Tyrrrz) [inline src/Sentry/Protocol/SentryTraceHeader.cs:61]: This facilitates this type of workflow:

```csharp
using var client = new HttpClient();
using var request = new HttpRequestMessage(HttpMethod.Get, "https://example.com");
SentrySdk.GetTraceHeader().Inject(request);
```

Could make it as extension methods instead on `HttpHeaders`/`HttpRequestMessage`/`HttpClient` instead but not sure if worth.

File: src/Sentry/Internal/Hub.cs (hunk 1 above)
Comment (bruno-garcia) [reply src/Sentry/Internal/Hub.cs:164]: I believe it needs to read the current span. In a world where we don't bind a whole transaction to the scope (also because it isnt' thread safe) we would only have a tree (in a list form) of spans. We could build this SentryTrace by getting the current span and with that the traceid and sampling decision. Isn't this spec'ed out? How is it in Java and Python?

/cc @marandaneto, @maciejwalkowiak, @untitaker

File: src/Sentry/Protocol/SentryTraceHeader.cs (hunk 3 above)
Comment (bruno-garcia) [reply src/Sentry/Protocol/SentryTraceHeader.cs:61]: I'm hoping we'll be able to run this through an integration instead of users having to do this manually on their `HttpClient`s. Maybe as a `MessageHandler` or even through ListenerHandler if that's possible:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/2bc7924b065ddef8199bf69c2456724980522557/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpHandlerDiagnosticListener.cs

https://github.com/open-telemetry/opentelemetry-dotnet/blob/2bc7924b065ddef8199bf69c2456724980522557/src/OpenTelemetry.Instrumentation.Http/Implementation/HttpInstrumentationEventSource.cs

File: src/Sentry/Internal/Hub.cs (hunk 1 above)
Comment (untitaker) [reply src/Sentry/Internal/Hub.cs:164]: Python assembles the entire trace tree all the time. Regardless of sampling decision it propagates a trace header with outbound HTTP requests that is derived from the current span.

If you feel strongly about performance and don't want to create span IDs when the transaction sampling decision is 0, I think you could take something other than the current span as long as the trace ID is propagated properly.

>In a world where we don't bind a whole transaction to the scope (also because it isnt' thread safe)

This already deviates from Python behavior so fundamentally that I cannot give any advice here. Python scopes are basically threadsafe and the current span is always bound to the scope. This is useful because it means outgoing error events can grab the current span from the scope to attach a trace context that makes sense.

Comment (bruno-garcia) [reply src/Sentry/Internal/Hub.cs:164]: Yeah this "Transaction on the Scope" Seem to have originated on JS, then the Spec and now Java and .NET.

The topic wasn't about avoiding allocations in this case but really that, the transaction on the scope (as opposed to only spans).

Comment (Tyrrrz) [reply src/Sentry/Internal/Hub.cs:164]: @bruno-garcia all of that makes sense, I was more interested in the `current span` part. Considering it doesn't store the transaction anywhere when you call `StartTransaction`, where would it get the current span?

File: src/Sentry/Protocol/SentryTraceHeader.cs (hunk 3 above)
Comment (Tyrrrz) [reply src/Sentry/Protocol/SentryTraceHeader.cs:61]: These would also be useful from inside the handlers, I believe. I guess I could also add another method `GetHttpHandler()` that wraps an existing handler to make it easier for those use cases. I'll look into `ListenerHandler`, first time hearing about it.

File: src/Sentry/Internal/Hub.cs (hunk 1 above)
Comment (bruno-garcia) [reply src/Sentry/Internal/Hub.cs:164]: If you want to make the transaction globally accessibly you must bind the transaction (or ideally the span list in a thread-ssafe manner instead, since we will have huge headaches with the Transaction thing on the scope!) to the Scope. It's optional. If it's not bound, it won't be available for integrations unless you programatically pass the value it's needed.

var tran = hub.StartTransction(..);

DoSomething("arg", tran);

void DoSomething(string arg, ITransaction tran)
{
 var client = new HttpClient(new SentryHandler(tran));
 client.Get()..
 // or
  var clietn = new HttpClient();
  var req = new HttpRequest();
  req.Headers.Add(tran.GetSentryTrace());
  client.Request(req);
}

Comment (Tyrrrz) [reply src/Sentry/Internal/Hub.cs:164]: Ok I see. So the code is right, it's just that it will rely on the user to set the transaction themselves.

**Comment by Tyrrrz** [comment]
@bruno-garcia should be final, please do a final review

File: src/Sentry/SentryClient.cs [hunk 4]
```diff
@@ -129,6 +129,17 @@ public void CaptureTransaction(ITransaction transaction)
                 return;
             }
 
+            if (!transaction.IsFinished)
+            {
+                _options.DiagnosticLogger?.LogWarning(
```
Comment (bruno-garcia) [inline src/Sentry/SentryClient.cs:134]: Loggin a warning makes sense.

Should we just finish the transaction and capture it instead?

How does it work in the other SDKs? Could be a behavior we align.

File: src/Sentry/SentryClient.cs [hunk 5]
```diff
@@ -129,6 +129,17 @@ public void CaptureTransaction(ITransaction transaction)
...
+                    "Transaction dropped because it's not finished. " +
+                    "Don't call hub.CaptureTransaction(...) directly. " +
```
Comment (bruno-garcia) [inline src/Sentry/SentryClient.cs:136]:
```suggestion
                    "You don't need to call hub.CaptureTransaction(...) directly. " +
```

File: src/Sentry/SentryHttpMessageHandler.cs [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+using System.Threading.Tasks;
+using Sentry.Extensibility;
+using Sentry.Protocol;
+
+namespace Sentry
+{
+    /// <summary>
+    /// Special HTTP message handler that can be used to propagate Sentry headers and other contextual information.
+    /// </summary>
+    public class SentryHttpMessageHandler : HttpMessageHandler
```
Comment (bruno-garcia) [inline src/Sentry/SentryHttpMessageHandler.cs:12]: `DelegatingHandler` isn't a good option?

File: test/Sentry.Tests/SentryHttpMessageHandlerTests.cs [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+using Sentry.Protocol;
+using Sentry.Testing;
+using Xunit;
+
+namespace Sentry.Tests
+{
+    public class SentryHttpMessageHandlerTests
+    {
+        [Fact]
+        public async Task SendAsync_SentryTraceHeaderNotSet_SetsHeader()
```
Comment (bruno-garcia) [inline test/Sentry.Tests/SentryHttpMessageHandlerTests.cs:15]: 1 test can really cover all edge cases this code can have?

**Review by bruno-garcia — APPROVED** [review]
I left some notes but it's good to merge. We can iterate with the other features etc.

File: src/Sentry/SentryClient.cs (hunk 4 above)
Comment (marandaneto) [reply src/Sentry/SentryClient.cs:134]: that's a good point, we don't check it at all on Java

Comment (marandaneto) [reply src/Sentry/SentryClient.cs:134]: what's about logging it as `INFO` instead but finishing and capturing it?

Comment (marandaneto) [reply src/Sentry/SentryClient.cs:134]: Ive raised this issue https://github.com/getsentry/sentry-java/issues/1208

File: test/Sentry.Tests/SentryHttpMessageHandlerTests.cs (hunk 7 above)
Comment (Tyrrrz) [reply test/Sentry.Tests/SentryHttpMessageHandlerTests.cs:15]: 2 tests for 2 branches, so I can't think of anything else ;)

File: src/Sentry/SentryClient.cs (hunk 4 above)
Comment (Tyrrrz) [reply src/Sentry/SentryClient.cs:137]: Changed it so that it captures with a warning.
